source/base/data_out_base.cc

```diff
@@ -5415,13 +5415,24 @@ namespace DataOutBase
                   {
                     vtk_data[0][0] = data_vectors(first_component, n);
                   }
-                else if ((size == 4) || (size == 9))
-                  // 2D, 4 elements or 3D 9 elements
+                else if (size == 4)
+                  // 2D, 4 elements
                   {
                     for (unsigned int c = 0; c < size; ++c)
                       {
                         const auto ind =
-                          Tensor<2, dim>::unrolled_to_component_indices(c);
+                          Tensor<2, 2>::unrolled_to_component_indices(c);
+                        vtk_data[ind[0]][ind[1]] =
+                          data_vectors(first_component + c, n);
```

Comment:
Yes, this looks right. You're just going to populate the top 2x2 sub-block.

I don't think the previous code was wrong, though. It relied on the fact that size==4 in 2d and size==9 whenever we got an output patch in 3d. But this is cleaner.